### ISSUE TITLE
[BUG] Offset trajectory with dump_modify skips a step if independent undump is used

### ISSUE THREAD

**Issue by jaclark5** [lammps/lammps#3647]
**Summary**

Background
Let's say I'm interested in a property calculated from two frames that are 10 time steps apart. Instead of writing out a dump file every 10 steps, I can dump a frame every 500 steps (dump 1), and then use dump_modify to dump a parallel trajectory with a 10 ts offset to dump 1 (dump 2). Let's say I run this for 20,000 ts. This works fine.

The Issue
If I add an additional dump every 5 ts for 5,000 ts to run at the same time (dump 3, let's say to calculate a residence time), then when I `undump 3`, dump 2 skips a step so my offset is 5010 for the remaining 19,000 ts for dump 1 and dump 2.

**LAMMPS Version and Platform**

LAMMPS (8 Feb 2023) on MacOS 13.2, Intel i5
LAMMPS (23 Jun 2022 - Update 2) CentOS 7.9.2009 x86_64

**Expected Behavior**

This is a sample to explain the issue for this section, the full input script is shown below.
```
dump 1 all custom ${dwfreq} nvt_${dwfreq}.lammpstrj id mol type q x y z ix iy iz
dump 2 all custom ${dwfreq} nvt_${dwfreq}_dw.lammpstrj id mol type q x y z ix iy iz
variable tdw equal stride(0,v_dwmaxts,${dwfreq})+${offset}
dump_modify   2 every v_tdw first yes
```

I expect the following time steps to be written to the dump files, regardless of whether another dump file is present or not:

ts_dump1   ts_ dump2
          0            0
      500        510
     1000       1010
          ...            ...
     4500       4510
     5000       **5010**
     5500       5510
          ...            ...

**Actual Behavior**

If another dump file is started and also stopped while dump 2 is running then a step is skipped when the run is reinitiated.

Resulting output:
ts_dump1   ts_ dump2
          0            0
      500        510
     1000       1010
          ...            ...
     4500       4510
     5000       **5510**
     5500       6010
          ...            ...

**Steps to Reproduce**

```
#
units           lj
atom_style      full

region          box block 0 10 0 10 0 10
create_box      2 box
create_atoms    1 single 5 5 5.5
create_atoms    1 single 5 5 4.5

pair_style      lj/cut 4.5
pair_coeff      * * 1.0 1.0
mass            * 1.0

###########

minimize 1.0e-4 1.0e-6 1000 10000
variable TK equal 2
velocity all create ${TK} 1234 rot yes dist gaussian
fix             1 all nvt temp ${TK} ${TK} 0.05

timestep        0.01
reset_timestep 0
write_data initial.data

##########

variable dwfreq equal 500
variable dwmaxts equal 20000
variable offset equal 10

dump 1 all custom ${dwfreq} nvt_${dwfreq}.lammpstrj id mol type q x y z ix iy iz
dump 2 all custom ${dwfreq} nvt_${dwfreq}_dw.lammpstrj id mol type q x y z ix iy iz
variable tdw equal stride(0,v_dwmaxts,${dwfreq})+${offset}
dump_modify   2 every v_tdw first yes

# !!! Comment this out to see that the issue is "resolved" !!!
dump 3 all custom 5 nvt_5.lammpstrj id mol type q x y z ix iy iz
run 5000
undump 3

variable nsteps equal ${dwmaxts}-5000
run  ${nsteps}
undump 1
undump 2
```

```python
import numpy as np
import matplotlib.pyplot as plt

def get_ts(filename):
    ts_array = []
    with open(filename, "r") as f:
        for line in f:
            if "TIMESTEP" in line:
                line = next(f).strip()
                ts_array.append(int(line))
    return np.array(ts_array)

freq = 500
data1 = get_ts("nvt_{}.lammpstrj".format(freq))
data2 = get_ts("nvt_{}_dw.lammpstrj".format(freq))

output = np.concatenate(
    (data1[:len(data2)][:,None], data2[:,None], (data2-data1[:len(data2)])[:,None]),
    axis=1
)
np.savetxt("timesteps.csv", output, delimiter=",")

plt.plot((data2-data1[:len(data2)])[1:])
plt.ylabel("Offset Between Dump1 and Dump2")
plt.savefig("plot.png",dpi=300)
```

**Comment by akohlmey** [comment]
Thanks for your report. I will take a look.

BTW: 
Please note that you could implement the type of output pattern you are looking for with a single dump command if you use something like this:
```
variable next equal step+(step%500==0)*10+(step%500!=0)*490
dump 1 all atom 1 test.dump
dump_modify 1 every v_next first yes
```

**Comment by akohlmey** [comment]
@jaclark5 I've looked closely at your input and have to conclude that LAMMPS is acting according to its documentation.
The problem is that your method of determining the timestep for the next dump is not compatible with breaking a run down into multiple parts the way you currently do. It has *nothing* to do with the additional dump command. You can comment that out and you still have the issue that you describe for as long as you keep the "run 5000" command.

The problem you have stems from the logic that LAMMPS applies when to do the next dump iteration. This is best illustrated by making both dumps use a variable:
```
variable dwfreq equal 500
variable dwmaxts equal 20000
variable offset equal 10

dump 1 all custom ${dwfreq} nvt_${dwfreq}.lammpstrj id mol type q x y z ix iy iz
dump 2 all custom ${dwfreq} nvt_${dwfreq}_dw.lammpstrj id mol type q x y z ix iy iz
variable tdw equal stride(0,v_dwmaxts,${dwfreq})+${offset}
dump_modify   2 every v_tdw first yes
variable t equal stride(0,v_dwmaxts,${dwfreq})
dump_modify  1 every v_t first yes
run 5000
print "t: ${t} tdw: ${tdw}"
```

As you can see from the print command output. The next step to output on dump 2 is computed to be step 5510. Step 5010 was scheduled during the "run 5000" command but was never reached. The next run starts at step 5000 and thus the stride() function outputs 5500 and adding the offset gives 5510.

If you change "run 5000" to "run 4990", the stride() function will start at 5000 since the second run starts at step 4990 and then the missed dump frame will be present.

In conclusion, the issue is that the stride() function does not provide the sequence of numbers that you are expecting, but does function exactly as documented. Thus the problem is in your input and this is not a bug in LAMMPS.

**Comment by jaclark5** [comment]
Thanks for your explanation, and sorry for the false alarm. I was able to achieve the result I wanted with with following (I want two files, otherwise your other suggestion would be great):
```
variable tdw equal (floor((step-${offset})/${dwfreq})+1)*${dwfreq}+${offset}
dump_modify   2 every v_tdw delay ${dwfreq} first yes
```
Use of the `delay` keyword overrides `first yes` in dump_modify though
